Egress IP fixes

alexanderConstantinescu opened the pull request:
<!--
Please make sure you've read and understood our contributing guidelines;
https://github.com/ovn-org/ovn-kubernetes/blob/master/CONTRIBUTING.md

** Make sure all your commits include a signature generated with `git commit -s` **

If this is a bug fix, make sure your description includes "fixes #xxxx", or
"closes #xxxx"

Trivial changes are exempt from following this template.
If your change is non-trivial, please provide the following information:
-->

**- What this PR does and why is it needed**

This PR fixes a couple of issues with the egress IP feature that merged yesterday with PR:  https://github.com/ovn-org/ovn-kubernetes/pull/1484. There is no real rush to get this in today and it can wait until next week. I didn't have time to do this yesterday with all the things going on.

Problems solved:

- When deleting the pod's egress setup (upon a pod delete) we were retrieving the IP from the OVN database and subsequently racing with `deleteLogicalPort`, in case we lost: we never cleaned up. I thus changed that to only look at the `pod.Status` instead, this resolves the race condition. Since a pod IP can't change we thus never really need to look in the OVN DB for that. This also simplified the code as I didn't need to pull in `ovnNBClient` into the data containers intended at managing egress IP assignments. 

- I never properly removed any watch handlers from the namespace / pod handlers used for egress IP assignment. This lead to very strange corner cases which have now been fixed. I've also added tests taking into account those cases.    

<!--
Write a short (one line) summary that describes the changes in this
pull request for inclusion in the changelog:
-->

alexanderConstantinescu commented:
/retest

alexanderConstantinescu commented:
/retest

danwinship commented on go-controller/pkg/ovn/egressip.go:
```diff
@@ -384,16 +382,23 @@ func (e *egressIPMode) getGatewayRouterJoinIP(node string, wantsIPv6 bool) (net.
-	_, podIps, err := util.GetPortAddresses(getPodKey(pod), e.ovnNBClient)
-	if err != nil || len(podIps) == 0 {
+func (e *egressIPMode) getPodIPs(pod *kapi.Pod) []net.IP {
+	if len(pod.Status.PodIPs) == 0 && pod.Status.PodIP == "" {
```
You don't need to look at `pod.Status.PodIP` (here or below (or anywhere else)). Unless you're in a cluster much older than anything we support, `PodIPs` will always include `PodIP`.

(Oh, maybe some of the unit tests that manually construct pods get this wrong? If so, fix the test case.)

danwinship commented on go-controller/pkg/ovn/egressip.go:
```diff
@@ -384,16 +382,23 @@ func (e *egressIPMode) getGatewayRouterJoinIP(node string, wantsIPv6 bool) (net.
+	if len(pod.Status.PodIPs) == 0 && pod.Status.PodIP == "" {
 		e.podRetry.Store(getPodKey(pod), true)
-		return nil, fmt.Errorf("pod does not have IP assigned, err: %v", err)
+		klog.V(5).Info("Pod does not have an IP assigned, retrying until it gets one")
```
except this function doesn't do any retry, and in the delete case (eg, if a pod is created and then deleted before CNI succeeds and so it never gets a pod IP) it wouldn't be retried so the message is wrong. So don't log anything here. You can log it in the caller if you need a log message

danwinship commented on go-controller/pkg/ovn/egressip_local_test.go:
```diff
@@ -106,12 +106,12 @@ var _ = Describe("Local gateway mode EgressIP Operations with", func() {
 					&ovntest.ExpectedCmd{
 						Cmd:    fmt.Sprintf("ovn-nbctl --timeout=15 --data=bare --format=table --no-heading --columns=options find logical_router name=GR_%s options:lb_force_snat_ip!=-", node2.name),
-						Output: fmt.Sprintf("lb_force_snat_ip=%s", nodeGatewayRouterIP),
+						Output: fmt.Sprintf("lb_force_snat_ip=%s", nodeLogicalRouterIPv6),
```
unrelated fix? I guess you were just using the wrong IP before and this just makes the test more correct but isn't actually fixing a bug?

alexanderConstantinescu replied:
You are right. I've split this out into its own commit. I'm keeping the commit in this PR however as the test I added, checking the handler deletion, benefits from it.

alexanderConstantinescu commented:
@danwinship : I've addressed your comments. PTAL

alexanderConstantinescu commented:
I have update the PR with the correct egress IP implementation for shared gateway mode. PR: https://github.com/ovn-org/ovn-kubernetes/pull/1616 will now bring in the necessary fix in OVN. So we should be good to go.

alexanderConstantinescu commented:
/hold 

I actually need another fix which I am currently implementing. Then this should be ready.

alexanderConstantinescu commented:
/hold cancel 

Added two commits fixing corner cases in the egress IP logic.

danwinship commented on go-controller/pkg/ovn/egressip.go:
```diff
@@ -416,55 +416,94 @@ func (e *egressIPMode) needsRetry(pod *kapi.Pod) bool {
+		if err != nil {
+			return fmt.Errorf("unable to find logical router policy: %s, stderr: %s, err: %v", status.EgressIP, stderr, err)
+		}
+		for _, policyID := range strings.Split(policyIDs, "\n") {
```
If there could actually be multiple rows then you need to split by `"\n\n"`, not `"\n"`. (Because that's how `ovn-*ctl find` works.) But there can only be one row output here anyway, right?

alexanderConstantinescu replied:
The `create`/`add` commands do not fail if the command is duplicated. So on restart we might risk adding the same policy twice, which is why I added this. It might be better checking that we never add something that already exists, but I figured that makes add one OVN transaction more costly.

alexanderConstantinescu replied:
FYI: even though the `create`/`add` commands lead to duplicated OVN objects, they do not seem to result in duplicated openflows:

```
ovn-nbctl --db tcp:172.21.0.2:6641 lr-policy-list ovn_cluster_router
Routing Policies
      1005 ip4.src == 10.244.0.2 && ip4.dst == 172.21.0.3 /* ovn-worker */         reroute               169.254.0.1
      1005 ip4.src == 10.244.1.2 && ip4.dst == 172.21.0.2 /* ovn-control-plane */         reroute               169.254.0.1
      1005 ip4.src == 10.244.2.2 && ip4.dst == 172.21.0.4 /* ovn-worker2 */         reroute               169.254.0.1
      1004 inport == "rtos-ovn-control-plane" && ip4.dst == 172.21.0.2 /* ovn-control-plane */         reroute                10.244.1.2
      1004 inport == "rtos-ovn-worker" && ip4.dst == 172.21.0.3 /* ovn-worker */         reroute                10.244.0.2
      1004 inport == "rtos-ovn-worker2" && ip4.dst == 172.21.0.4 /* ovn-worker2 */         reroute                10.244.2.2
       101 ip4.src == 10.244.0.0/16 && ip4.dst == 10.244.0.0/16           allow
       101 ip4.src == 10.244.0.0/16 && ip4.dst == 172.21.0.2/32           allow
       101 ip4.src == 10.244.0.0/16 && ip4.dst == 172.21.0.3/32           allow
       101 ip4.src == 10.244.0.0/16 && ip4.dst == 172.21.0.4/32           allow
       100      ip4.src == 10.244.0.3 && ip4.dst == 0.0.0.0/0         reroute                100.64.2.1
       100      ip4.src == 10.244.0.3 && ip4.dst == 0.0.0.0/0         reroute                100.64.1.1
       100      ip4.src == 10.244.0.3 && ip4.dst == 0.0.0.0/0         reroute                100.64.1.1
       100      ip4.src == 10.244.2.3 && ip4.dst == 0.0.0.0/0         reroute                100.64.1.1

[root@ovn-control-plane ~]# ovs-ofctl -O OpenFlow13 dump-flows br-int | grep 10.244.0.3
 cookie=0xa421b605, duration=198.260s, table=20, n_packets=0, n_bytes=0, priority=100,ip,metadata=0x1,nw_src=10.244.0.3 actions=load:0x64400101->NXM_NX_XXREG0[96..127],load:0x64400102->NXM_NX_XXREG0[64..95],set_field:0a:58:64:40:01:02->eth_src,set_field:0x7->reg15,load:0x1->NXM_NX_REG10[0],resubmit(,21)
 cookie=0x3cd483df, duration=22978.596s, table=21, n_packets=0, n_bytes=0, priority=50,arp,metadata=0x4,arp_tpa=10.244.0.3,arp_op=1 actions=move:NXM_OF_ETH_SRC[]->NXM_OF_ETH_DST[],set_field:0a:58:0a:f4:00:03->eth_src,set_field:2->arp_op,move:NXM_NX_ARP_SHA[]->NXM_NX_ARP_THA[],set_field:0a:58:0a:f4:00:03->arp_sha,move:NXM_OF_ARP_SPA[]->NXM_OF_ARP_TPA[],set_field:10.244.0.3->arp_spa,move:NXM_NX_REG14[]->NXM_NX_REG15[],load:0x1->NXM_NX_REG10[0],resubmit(,32)

``` 

I am thus unsure if it's worth running a `find` command before each `create` command, when setting up egress for pods.

alexanderConstantinescu replied:
OK, circling around on this: I am adding `--format=csv` to avoid the `\n\n`. Concerning the duplicated logical router policies / nat rules on restart. I don't see any duplicated openflows, even though the OVN objects are duplicated. I will need to confirm this tomorrow with @numansiddique , i.e: that having the exact same `Logical_router_policy` and `NAT` object twice does not duplicate the openflows.

danwinship replied:
> I don't see any duplicated openflows, even though the OVN objects are duplicated.

Yeah, it's a no-op to re-add a flow that already exists. (It neither duplicates the flow nor returns an error.)

```
[root@p50 ovn-kubernetes]# ovs-ofctl add-flow br0 table=0,priority=100,ip,actions=output:0
[root@p50 ovn-kubernetes]# ovs-ofctl add-flow br0 table=0,priority=100,ip,actions=output:0
[root@p50 ovn-kubernetes]# echo $?
0
[root@p50 ovn-kubernetes]# ovs-ofctl dump-flows br0
 cookie=0x0, duration=1.334s, table=0, n_packets=0, n_bytes=0, priority=100,ip actions=output:0
[root@p50 ovn-kubernetes]# 
```

alexanderConstantinescu replied:
So are you fine with not performing an additional `find` before doing a `create`, and having the delete remove multiple IDs (should they exist), @danwinship ?

danwinship replied:
> The `create`/`add` commands do not fail if the command is duplicated. So on restart we might risk adding the same policy twice, which is why I added this.

Huh? This is the delete code, not the add code...

We should not re-add the same OVN object on every restart, even if it doesn't result in duplicate OpenFlow. If we have to do a `find`-then-`create` to ensure that, then we should do that

alexanderConstantinescu replied:
> Huh? This is the delete code, not the add code...

Yeah, I was explaining why the delete code looked the way it did, which was because of the add code. 

I will make sure we `find`-then-`create` then.

danwinship commented on go-controller/pkg/ovn/egressip_shared.go:
```diff
@@ -42,34 +37,70 @@ func (e *egressIPShared) deletePodEgressIP(eIP *egressipv1.EgressIP, pod *kapi.P
-				return fmt.Errorf("OVN transaction error, stderr: %s, err: %v", stderr, err)
+				return fmt.Errorf("unable to find nat ID, stderr: %s, err: %v", stderr, err)
+			}
+			for _, natID := range strings.Split(natIDs, "\n") {
```
likewise

danwinship commented on go-controller/pkg/ovn/egressip.go:
```diff
@@ -416,55 +416,94 @@ func (e *egressIPMode) needsRetry(pod *kapi.Pod) bool {
+				"logical_router_policy",
+				"action=reroute",
+				fmt.Sprintf("match=\"%s\"", filterOption),
+				fmt.Sprintf("priority=%v", egressIPRereoutePriority),
```
the constant name is misspelled ("rerEoute")

alexanderConstantinescu commented on go-controller/pkg/ovn/egressip.go:
```diff
@@ -246,8 +246,8 @@ func (oc *Controller) deleteNamespacePodsEgressIP(eIP *egressipv1.EgressIP, name
 func (oc *Controller) assignEgressIPs(eIP *egressipv1.EgressIP) error {
-	assignments := []egressipv1.EgressIPStatusItem{}
 	oc.eIPAllocatorMutex.Lock()
+	assignments := []egressipv1.EgressIPStatusItem{}
```
Sure, it might be that I misinterpreted the data race report.

```
WARNING: DATA RACE
Read at 0x00c0002f9830 by goroutine 26:
  github.com/ovn-org/ovn-kubernetes/go-controller/pkg/crd/egressip/v1.(*EgressIP).DeepCopyInto()
      /home/aconstan/go/src/github.com/ovn-org/ovn-kubernetes/go-controller/pkg/crd/egressip/v1/zz_generated.deepcopy.go:29 +0x65
  github.com/ovn-org/ovn-kubernetes/go-controller/pkg/crd/egressip/v1.(*EgressIP).DeepCopyObject()
      /home/aconstan/go/src/github.com/ovn-org/ovn-kubernetes/go-controller/pkg/crd/egressip/v1/zz_generated.deepcopy.go:43 +0xa8
  k8s.io/client-go/testing.(*tracker).Get()
      /home/aconstan/go/src/github.com/ovn-org/ovn-kubernetes/go-controller/vendor/k8s.io/client-go/testing/fixture.go:309 +0x33d
  k8s.io/client-go/testing.ObjectReaction.func1()
      /home/aconstan/go/src/github.com/ovn-org/ovn-kubernetes/go-controller/vendor/k8s.io/client-go/testing/fixture.go:111 +0x45f
  k8s.io/client-go/testing.(*SimpleReactor).React()
      /home/aconstan/go/src/github.com/ovn-org/ovn-kubernetes/go-controller/vendor/k8s.io/client-go/testing/fixture.go:534 +0x66
  k8s.io/client-go/testing.(*Fake).Invokes()
      /home/aconstan/go/src/github.com/ovn-org/ovn-kubernetes/go-controller/vendor/k8s.io/client-go/testing/fake.go:141 +0x30d
  github.com/ovn-org/ovn-kubernetes/go-controller/pkg/crd/egressip/v1/apis/clientset/versioned/typed/egressip/v1/fake.(*FakeEgressIPs).Create()
      /home/aconstan/go/src/github.com/ovn-org/ovn-kubernetes/go-controller/pkg/crd/egressip/v1/apis/clientset/versioned/typed/egressip/v1/fake/fake_egressip.go:82 +0x237
  github.com/ovn-org/ovn-kubernetes/go-controller/pkg/ovn.glob..func4.5.3.1()
      /home/aconstan/go/src/github.com/ovn-org/ovn-kubernetes/go-controller/pkg/ovn/egressip_local_test.go:278 +0xf51
  github.com/urfave/cli/v2.(*App).RunContext()
      /home/aconstan/go/src/github.com/ovn-org/ovn-kubernetes/go-controller/vendor/github.com/urfave/cli/v2/app.go:315 +0xaa9
  github.com/ovn-org/ovn-kubernetes/go-controller/pkg/ovn.glob..func4.5.3()
      /home/aconstan/go/src/github.com/ovn-org/ovn-kubernetes/go-controller/vendor/github.com/urfave/cli/v2/app.go:215 +0x2c7
  github.com/onsi/ginkgo/internal/leafnodes.(*runner).runSync()
      /home/aconstan/go/src/github.com/ovn-org/ovn-kubernetes/go-controller/vendor/github.com/onsi/ginkgo/internal/leafnodes/runner.go:113 +0xdf
  github.com/onsi/ginkgo/internal/leafnodes.(*runner).run()
      /home/aconstan/go/src/github.com/ovn-org/ovn-kubernetes/go-controller/vendor/github.com/onsi/ginkgo/internal/leafnodes/runner.go:64 +0x107
  github.com/onsi/ginkgo/internal/leafnodes.(*ItNode).Run()
      /home/aconstan/go/src/github.com/ovn-org/ovn-kubernetes/go-controller/vendor/github.com/onsi/ginkgo/internal/leafnodes/it_node.go:26 +0x87
  github.com/onsi/ginkgo/internal/spec.(*Spec).runSample()
      /home/aconstan/go/src/github.com/ovn-org/ovn-kubernetes/go-controller/vendor/github.com/onsi/ginkgo/internal/spec/spec.go:215 +0x6e1
  github.com/onsi/ginkgo/internal/spec.(*Spec).Run()
      /home/aconstan/go/src/github.com/ovn-org/ovn-kubernetes/go-controller/vendor/github.com/onsi/ginkgo/internal/spec/spec.go:138 +0x167
  github.com/onsi/ginkgo/internal/specrunner.(*SpecRunner).runSpec()
      /home/aconstan/go/src/github.com/ovn-org/ovn-kubernetes/go-controller/vendor/github.com/onsi/ginkgo/internal/specrunner/spec_runner.go:200 +0x172
  github.com/onsi/ginkgo/internal/specrunner.(*SpecRunner).runSpecs()
      /home/aconstan/go/src/github.com/ovn-org/ovn-kubernetes/go-controller/vendor/github.com/onsi/ginkgo/internal/specrunner/spec_runner.go:170 +0x216
  github.com/onsi/ginkgo/internal/specrunner.(*SpecRunner).Run()
      /home/aconstan/go/src/github.com/ovn-org/ovn-kubernetes/go-controller/vendor/github.com/onsi/ginkgo/internal/specrunner/spec_runner.go:66 +0x140
  github.com/onsi/ginkgo/internal/suite.(*Suite).Run()
      /home/aconstan/go/src/github.com/ovn-org/ovn-kubernetes/go-controller/vendor/github.com/onsi/ginkgo/internal/suite/suite.go:62 +0x69f
  github.com/onsi/ginkgo.RunSpecsWithCustomReporters()
      /home/aconstan/go/src/github.com/ovn-org/ovn-kubernetes/go-controller/vendor/github.com/onsi/ginkgo/ginkgo_dsl.go:226 +0x338
  github.com/onsi/ginkgo.RunSpecsWithDefaultAndCustomReporters()
      /home/aconstan/go/src/github.com/ovn-org/ovn-kubernetes/go-controller/vendor/github.com/onsi/ginkgo/ginkgo_dsl.go:214 +0x11f
  github.com/onsi/ginkgo.RunSpecs()
      /home/aconstan/go/src/github.com/ovn-org/ovn-kubernetes/go-controller/vendor/github.com/onsi/ginkgo/ginkgo_dsl.go:205 +0x1a2
  github.com/ovn-org/ovn-kubernetes/go-controller/pkg/ovn.TestClusterNode()
      /home/aconstan/go/src/github.com/ovn-org/ovn-kubernetes/go-controller/pkg/ovn/ovn_suite_test.go:12 +0xf8
  testing.tRunner()
      /usr/lib/golang/src/testing/testing.go:909 +0x199

Previous write at 0x00c0002f9830 by goroutine 192:
  github.com/ovn-org/ovn-kubernetes/go-controller/pkg/ovn.(*Controller).assignEgressIPs.func1()
      /home/aconstan/go/src/github.com/ovn-org/ovn-kubernetes/go-controller/pkg/ovn/egressip.go:252 +0x77
  github.com/ovn-org/ovn-kubernetes/go-controller/pkg/ovn.(*Controller).assignEgressIPs()
      /home/aconstan/go/src/github.com/ovn-org/ovn-kubernetes/go-controller/pkg/ovn/egressip.go:309 +0xdeb
  github.com/ovn-org/ovn-kubernetes/go-controller/pkg/ovn.(*Controller).addEgressIP()
      /home/aconstan/go/src/github.com/ovn-org/ovn-kubernetes/go-controller/pkg/ovn/egressip.go:38 +0x70d
  github.com/ovn-org/ovn-kubernetes/go-controller/pkg/ovn.(*Controller).WatchEgressIP.func1()
      /home/aconstan/go/src/github.com/ovn-org/ovn-kubernetes/go-controller/pkg/ovn/ovn.go:824 +0x80
  k8s.io/client-go/tools/cache.(*ResourceEventHandlerFuncs).OnAdd()
      /home/aconstan/go/src/github.com/ovn-org/ovn-kubernetes/go-controller/vendor/k8s.io/client-go/tools/cache/controller.go:218 +0x8c
  k8s.io/client-go/tools/cache.FilteringResourceEventHandler.OnAdd()
      /home/aconstan/go/src/github.com/ovn-org/ovn-kubernetes/go-controller/vendor/k8s.io/client-go/tools/cache/controller.go:250 +0x7a
  github.com/ovn-org/ovn-kubernetes/go-controller/pkg/factory.(*Handler).OnAdd()
      /home/aconstan/go/src/github.com/ovn-org/ovn-kubernetes/go-controller/pkg/factory/factory.go:57 +0xa6
  github.com/ovn-org/ovn-kubernetes/go-controller/pkg/factory.(*informer).newFederatedHandler.func1.1()
      /home/aconstan/go/src/github.com/ovn-org/ovn-kubernetes/go-controller/pkg/factory/factory.go:263 +0x54
  github.com/ovn-org/ovn-kubernetes/go-controller/pkg/factory.(*informer).forEachHandler()
      /home/aconstan/go/src/github.com/ovn-org/ovn-kubernetes/go-controller/pkg/factory/factory.go:124 +0x1f0
  github.com/ovn-org/ovn-kubernetes/go-controller/pkg/factory.(*informer).newFederatedHandler.func1()
      /home/aconstan/go/src/github.com/ovn-org/ovn-kubernetes/go-controller/pkg/factory/factory.go:262 +0x8a
  k8s.io/client-go/tools/cache.(*ResourceEventHandlerFuncs).OnAdd()
      /home/aconstan/go/src/github.com/ovn-org/ovn-kubernetes/go-controller/vendor/k8s.io/client-go/tools/cache/controller.go:218 +0x8c
  k8s.io/client-go/tools/cache.(*processorListener).run.func1()
      /home/aconstan/go/src/github.com/ovn-org/ovn-kubernetes/go-controller/vendor/k8s.io/client-go/tools/cache/shared_informer.go:744 +0x3ad
  k8s.io/apimachinery/pkg/util/wait.BackoffUntil.func1()
      /home/aconstan/go/src/github.com/ovn-org/ovn-kubernetes/go-controller/vendor/k8s.io/apimachinery/pkg/util/wait/wait.go:155 +0x6f
  k8s.io/apimachinery/pkg/util/wait.BackoffUntil()
      /home/aconstan/go/src/github.com/ovn-org/ovn-kubernetes/go-controller/vendor/k8s.io/apimachinery/pkg/util/wait/wait.go:156 +0xb3
  k8s.io/apimachinery/pkg/util/wait.JitterUntil()
      /home/aconstan/go/src/github.com/ovn-org/ovn-kubernetes/go-controller/vendor/k8s.io/apimachinery/pkg/util/wait/wait.go:133 +0x10d
  k8s.io/client-go/tools/cache.(*processorListener).run()
      /home/aconstan/go/src/github.com/ovn-org/ovn-kubernetes/go-controller/vendor/k8s.io/apimachinery/pkg/util/wait/wait.go:90 +0xa2
  k8s.io/client-go/tools/cache.(*processorListener).run-fm()
      /home/aconstan/go/src/github.com/ovn-org/ovn-kubernetes/go-controller/vendor/k8s.io/client-go/tools/cache/shared_informer.go:732 +0x41
  k8s.io/apimachinery/pkg/util/wait.(*Group).Start.func1()
      /home/aconstan/go/src/github.com/ovn-org/ovn-kubernetes/go-controller/vendor/k8s.io/apimachinery/pkg/util/wait/wait.go:73 +0x6a

Goroutine 26 (running) created at:
  testing.(*T).Run()
      /usr/lib/golang/src/testing/testing.go:960 +0x651
  testing.runTests.func1()
      /usr/lib/golang/src/testing/testing.go:1202 +0xa6
  testing.tRunner()
      /usr/lib/golang/src/testing/testing.go:909 +0x199
  testing.runTests()
      /usr/lib/golang/src/testing/testing.go:1200 +0x521
  testing.(*M).Run()
      /usr/lib/golang/src/testing/testing.go:1117 +0x2ff
  main.main()
      _testmain.go:44 +0x223

Goroutine 192 (running) created at:
  k8s.io/apimachinery/pkg/util/wait.(*Group).Start()
      /home/aconstan/go/src/github.com/ovn-org/ovn-kubernetes/go-controller/vendor/k8s.io/apimachinery/pkg/util/wait/wait.go:71 +0x6f
  k8s.io/client-go/tools/cache.(*sharedProcessor).run.func1()
      /home/aconstan/go/src/github.com/ovn-org/ovn-kubernetes/go-controller/vendor/k8s.io/client-go/tools/cache/shared_informer.go:590 +0x1b4
  k8s.io/client-go/tools/cache.(*sharedProcessor).run()
      /home/aconstan/go/src/github.com/ovn-org/ovn-kubernetes/go-controller/vendor/k8s.io/client-go/tools/cache/shared_informer.go:594 +0x3f
  k8s.io/client-go/tools/cache.(*sharedProcessor).run-fm()
      /home/aconstan/go/src/github.com/ovn-org/ovn-kubernetes/go-controller/vendor/k8s.io/client-go/tools/cache/shared_informer.go:585 +0x4b
  k8s.io/apimachinery/pkg/util/wait.(*Group).StartWithChannel.func1()
      /home/aconstan/go/src/github.com/ovn-org/ovn-kubernetes/go-controller/vendor/k8s.io/apimachinery/pkg/util/wait/wait.go:56 +0x45
  k8s.io/apimachinery/pkg/util/wait.(*Group).Start.func1()
      /home/aconstan/go/src/github.com/ovn-org/ovn-kubernetes/go-controller/vendor/k8s.io/apimachinery/pkg/util/wait/wait.go:73 +0x6a
==================

```

alexanderConstantinescu replied:
Yeah, I definitely misinterpreted this race report. I just hit it again with my "fix". I don't understand why it's pointing to line `egressip.go:252` (effectively `eIP.Status.Items = assignments`) and `zz_generated.deepcopy.go:29`

alexanderConstantinescu replied:
OK, I think I _might_ understand it finally. Some tests are creating EgressIP objects, as:

```
_, err := fakeOvn.fakeEgressIPClient.K8sV1().EgressIPs().Create(context.TODO(), &eIP, metav1.CreateOptions{})
```

The hypothesis is that several go-routines are spawned for this when the race detector is set to "on". One of this go routines will be in the process of executing code in `pkg/crd/egressip/v1/apis/clientset/versioned/typed/egressip/v1` while another will be in `egressip.go` and L252 and modifying `eIP` by setting its status. I thus need to modify the tests to have this happen "safely" as I've done for `EgressIPs().Get` and `getEgressIPStatusSafely`, etc.

alexanderConstantinescu replied:
OK, seems fixed now with the additional patch I am working on. Will push soon.

trozet replied:
so it sounds like this was a problem in the test code and/or something resolved by an additional commit. So is this commit still needed?

alexanderConstantinescu replied:
> I think the problem here isn't really "thread-safety" per se; it's that you're modifying objects that you received from the informer cache

Gahh, of course! Thank you for that pointer, I removed all the locks I added in the tests and switched to modifying the copy and updating the API server with that.

> So is this commit still needed?

I squashed that commit. I still prefer to wrap all operations inside the lock however, so I kept the code.

alexanderConstantinescu commented on go-controller/pkg/ovn/egressip.go:
```diff
@@ -416,55 +416,94 @@ func (e *egressIPMode) needsRetry(pod *kapi.Pod) bool {
-			if err != nil && !strings.Contains(stderr, policyAlreadyExistsMsg) {
-				return fmt.Errorf("unable to create logical router policy for egress IP: %s, stderr: %s, err: %v", status.EgressIP, stderr, err)
-			}
+			filterOption = fmt.Sprintf("ip6.src == %s && ip6.dst == ::/0", podIP.String())
```
Hmm, in the beginning I added this as to avoid rerouting east<->west traffic, but at one moment I think I noticed that didn't work and thus added the no-reroute policies created by `createDefaultNoReroutePodPolicies` / `createDefaultNoRerouteNodePolicies`. So I think it's safe to remove this.

alexanderConstantinescu commented:
@danwinship , addressed all comments. PTAL, and let me know if you'd prefer I squash some commits or not. I've kept them separate for easier review / tracking of what has been said here.

alexanderConstantinescu commented:
> changes look good other than the race-detector problem which I commented on above.
> 
> Please squash the fixups back into their original commits in the next round

Updated and squashed. PTAL

danwinship commented on go-controller/pkg/ovn/egressip.go:
```diff
@@ -74,10 +74,14 @@ func (oc *Controller) deleteEgressIP(eIP *egressipv1.EgressIP) error {
 
 	oc.egressIPNamespaceHandlerMutex.Lock()
 	defer oc.egressIPNamespaceHandlerMutex.Unlock()
+	nH := oc.egressIPNamespaceHandlerCache[getEgressIPKey(eIP)]
+	oc.watchFactory.RemoveNamespaceHandler(&nH)
 	delete(oc.egressIPNamespaceHandlerCache, getEgressIPKey(eIP))
```
for safety:

```suggestion
	if nH, exists := oc.egressIPNamespaceHandlerCache[getEgressIPKey(eIP)]; exists {
		oc.watchFactory.RemoveNamespaceHandler(&nH)
		delete(oc.egressIPNamespaceHandlerCache, getEgressIPKey(eIP))
	}
```

likewise below

danwinship commented on go-controller/pkg/ovn/egressip.go:
```diff
@@ -412,61 +412,115 @@ func (e *egressIPMode) needsRetry(pod *kapi.Pod) bool {
+		if err != nil {
+			return err
+		}
+		for _, policyID := range strings.Split(policyIDs, "\n") {
```
It would make more sense to have `findLRPolicyIDs` do the split itself and return an array.

But also, it needs to check if the return value from `util.RunOVNNbctl` is `""` first and return `nil` in that case, since `strings.Split("", "\n")` will return `[]string{""}`, and so you would end up running "`ovn-nbctl remove logical_router ... policies ""`" here and then getting an error.

danwinship commented on go-controller/pkg/ovn/egressip_local.go:
```diff
@@ -10,6 +10,7 @@ import (
 const (
 	// In case we restart we need accept executing ovn-nbctl commands with this error.
+	// The ovn-nbctl API does not support `--may-exist` for `lr-policy-add`
 	policyAlreadyExistsMsg = "Same routing policy already existed"
```
This is only used from `egressip.go`, not `egressip_local.go` so it ought to be there?

danwinship commented on go-controller/pkg/ovn/egressip_shared.go:
```diff
@@ -46,36 +41,87 @@ func (e *egressIPShared) deletePodEgressIP(eIP *egressipv1.EgressIP, pod *kapi.P
 	return nil
 }
+
+func findNatIDs(egressIPName, podIP, egressIP string) (string, error) {
```
same comments as with `findLRPolicyIDs`

danwinship commented on go-controller/pkg/ovn/egressip_local_test.go:
```diff
@@ -25,17 +25,13 @@ var _ = Describe("Local gateway mode EgressIP Operations with", func() {
 	getEgressIPStatusLenSafely := func(egressIPName string) func() int {
 		return func() int {
-			fakeOvn.controller.eIPAllocatorMutex.Lock()
-			defer fakeOvn.controller.eIPAllocatorMutex.Unlock()
```
So do we need these `Safely` functions any more? (At least, the "Safely" is no longer meaningful is it?)

alexanderConstantinescu replied:
You're right, I was going to remove "Safely" yesterday, but forgot. The functions are handy though, given how prevalent tests are using this.

danwinship commented on go-controller/pkg/ovn/egressip.go:
```diff
@@ -534,12 +534,13 @@ func (oc *Controller) addEgressNode(egressNode *kapi.Node) error {
 		}
-		if err := oc.addEgressIP(eIP); err != nil {
-			klog.Errorf("Re-assignment for EgressIP: unable to assign EgressIP: %s, err: %v", eIP.Name, err)
+		eIPCopy := eIP.DeepCopy()
```
alternatively you could just do `eIP = eIP.DeepCopy()` since I don't think you use the original again. (likewise elsewhere)

alexanderConstantinescu commented:
@danwinship , just pushed again with a version addressing your latest comments.

alexanderConstantinescu commented:
Hmm, last build failure was caused by:

```
• Failure [5.003 seconds]
Hybrid SDN Master Operations
/home/runner/work/ovn-kubernetes/ovn-kubernetes/go-controller/hybrid-overlay/pkg/controller/master_test.go:70
  cleans up a Linux node when the OVN hostsubnet annotation is removed [It]
  /home/runner/work/ovn-kubernetes/ovn-kubernetes/go-controller/hybrid-overlay/pkg/controller/master_test.go:311

  Timed out after 5.000s.
  Expected
      <map[string]string | len:1>: {
          "k8s.ovn.org/hybrid-overlay-distributed-router-gateway-mac": "00:00:00:52:19:d2",
      }
  not to have key
      <string>: k8s.ovn.org/hybrid-overlay-distributed-router-gateway-mac

  /home/runner/work/ovn-kubernetes/ovn-kubernetes/go-controller/hybrid-overlay/pkg/controller/master_test.go:373
```

Seems like we have flakes (issues?) with the hybrid overlay code.
